### ISSUE TITLE
`mapreducedim` is the reason for backward pass slowness

### ISSUE THREAD

**Issue by jekbradbury** [avik-pal/DeepLearningBenchmarks#1]
The (known to be slow) GPU kernel for `mapreducedim` is the bottleneck for the backwards pass of convolution layers (the selected range is one forwards and backwards pass of the 3x3 conv from the benchmarks; compare `ptxcall_mapreducedim...` to `cudnn::detail::wgrad...`, the actual conv backwards call):

<img width="1392" alt="screen shot 2018-07-16 at 11 20 17 pm" src="https://user-images.githubusercontent.com/11729078/42785186-c88064ce-8951-11e8-9c65-687fb1dc9542.png">

This isn't called in the backwards pass of the convolution itself, but the backwards pass of the [broadcasted activation function that follows it](https://github.com/FluxML/Flux.jl/blob/v0.5.3/src/layers/conv.jl#L44):
```
Stacktrace:
 [1] _mapreducedim! at /root/.julia/v0.6/GPUArrays/src/mapreduce.jl:83 [inlined]
 [2] mapreducedim! at ./reducedim.jl:210 [inlined]
 [3] mapreducedim at ./reducedim.jl:242 [inlined]
 [4] sum at ./reducedim.jl:583 [inlined]
 [5] sum at ./reducedim.jl:585 [inlined]
 [6] unbroadcast(::TrackedArray{…,CuArray{Float32,4}}, ::CuArray{Float32,4}) at /root/.julia/v0.6/Flux/src/tracker/array.jl:392
 [7] macro expansion at /root/.julia/v0.6/Flux/src/tracker/back.jl:49 [inlined]
 [8] (::Flux.Tracker.##69#71)(::TrackedArray{…,CuArray{Float32,4}}, ::CuArray{Float32,4}) at /root/.julia/v0.6/Flux/src/tracker/array.jl:405
 [9] foreach(::Function, ::Tuple{TrackedArray{…,CuArray{Float32,4}},TrackedArray{…,CuArray{Float32,4}}}, ::Tuple{CuArray{Float32,4},CuArray{Float32,4}}, ::Vararg{Tuple{CuArray{Float32,4},CuArray{Float32,4}},N} where N) at ./abstractarray.jl:1734
 [10] back_(::Flux.Tracker.Call{Flux.Tracker.Broadcasted{Flux.##97#99{Base.#identity},CuArray{ForwardDiff.Dual{Void,Float32,2},4}},Tuple{TrackedArray{…,CuArray{Float32,4}},TrackedArray{…,CuArray{Float32,4}}}}, ::CuArray{Float32,4}, ::CuArray{Float32,4}) at /root/.julia/v0.6/Flux/src/tracker/back.jl:25
 [11] back(::Flux.Tracker.Tracked{CuArray{Float32,4}}, ::CuArray{Float32,4}) at /root/.julia/v0.6/Flux/src/tracker/back.jl:38
 [12] back!(::TrackedArray{…,CuArray{Float32,4}}, ::CuArray{Float32,4}) at /root/.julia/v0.6/Flux/src/tracker/back.jl:63
```

This is particularly unfortunate because the activation function here is just the identity, but in general it's not clear to me why taking the gradient of an elementwise operation should require a `sum` in the case where the operation doesn't actually extrude/broadcast any arrays. Any thoughts, @mikeinnes? Other than that I should finish/generalize the mapreducedim kernel I have so I can PR it to CuArrays? Is this fixed in the refactored AD API?

**Comment by avik-pal** [comment]
I think this also explains why `CuArrays.CUDNN.cudnnConvolutionForward($y, $x, $w, padding = 1)` runs in `140 microsec` while the total forward pass for Conv 3x3 runs in about `1 millisec`.

**Comment by jekbradbury** [comment]
I don’t think there’s a call to `mapreducedim` in the Conv forward pass; most of the overhead there in my benchmark looks like it comes from CUDA memory allocation and is probably going to be more difficult to deal with.

This is because the caching allocator strategy used by all dynamic DL frameworks (and Flux) works best when the host language uses reference counting that frees variables at exactly the moment they go out of scope; in Julia the runtime holds on to temporaries between iterations, forcing additional allocations. This is probably a lot better at the steady state (the above profile is for iteration 3) but it might still be worth solving. We’ll have to either a) explicitly trigger GC of temporaries after the end of each iteration, which I don’t know how to do other than by running `gc()`, or b) preallocate temporary buffers at JIT time and/or network construction time and explicitly reuse those buffers for each iteration.

We probably need to do (a) for now but (b) sounds much more fun.

**Comment by MikeInnes** [comment]
> it's not clear to me why taking the gradient of an elementwise operation should require a sum

Is this the `Conv` layer or the plain function? The layer has a bias which will get expanded. If not, it might be that we're calling `sum` with no dimensions to reduce – something we can fix easily.

> Other than that I should finish/generalize the mapreducedim kernel I have so I can PR it to CuArrays?

Please do! FWIW I'm totally fine with having fast kernels for special cases (2D etc) and falling back elsewhere. It's going to take some time to work out how to do really efficient ND kernels on the GPU, I think.

> trigger GC of temporaries after the end of each iteration

Might be worth throwing a `gc()` in as a callback to see what the impact is. If it's effective we can throw it into the Flux training loop.

On (b), there are some thoughts about doing some kind of conservative refcounting / dynamic escape analysis. I expect this is a long way off though.

**Comment by jekbradbury** [comment]
Mike is right—I missed that the bias is a true broadcasting operation; that makes a lot more sense than blaming the activation function.

But even so, PyTorch doesn’t need an explicit `sum` call anywhere in the convolution backward pass because they use `cudnnConvolutionBackwardBias` to compute the bias gradient. (It’s a little surprising, but kind of reasonable if you think about it, that in the forward pass you have to add the bias yourself, but cuDNN can still compute its gradient for you).

We should also make the layer-level comparison a little more fair by making sure the identity activation function doesn’t add even a single kernel, as PyTorch doesn’t integrate an activation function into their convolution layer.

**Comment by avik-pal** [comment]
> We should also make the layer-level comparison a little more fair by making sure the identity activation function doesn’t add even a single kernel, as PyTorch doesn’t integrate an activation function into their convolution layer.

I am not sure how to make this type of comparison given the existing definition of Conv we have. But for comparison purpose I can define a new struct of Conv which doesn't contain the activation function.

> PyTorch doesn’t need an explicit sum call anywhere in the convolution backward pass because they use `cudnnConvolutionBackwardBias` to compute the bias gradient.

If needed I think we can consider calling `cudnnConvolutionBiasActivationForward` directly instead of `cudnnConvolutionForward` and compute the bias grad using `cudnnConvolutionBackwardBias`

**Comment by MikeInnes** [comment]
If we are adding the bias ourselves, we need a broadcast op anyway and the identity activation function costs nothing. If we change to using cudnn to do that for us, we can just dispatch on the activation function (like the RNNs do).

It seems increasingly like we just need to hook into CUDNN at the level of the whole `Conv` layer, rather than just its components, but this should be pretty easy to do.

**Comment by jekbradbury** [comment]
I missed that second cuDNN forward function, Avik—that sounds like a good idea (and will help us get closer on the model benchmarks without more difficult changes to Flux)

**Comment by RoyiAvital** [comment]
Any chance testing `KNet.jl` as well?
